### DISCUSSION TITLE
reading the BRO data

### DISCUSSION THREAD

**Discussion by ArtemisRo** [ArtesiaWater/hydropandas#104] [General]
Hi every one,
I have downloaded some wells from the BROloket website but their format is .xml. I want to read them using the HydroPandas but actually I do not know how I should open these files. could you please let me know how to read them?

**Comment by dbrakenhoff** [comment]
Hi @rt84ro,

This is currently not possible because the XML reading code is embedded in the methods that download the BRO data directly. It should be fairly simple to separate the the downloading and parsing of XML results (given that the API and manual download produce the same files), but we haven´t done that yet. I'll create an issue to track this request. 

In the meantime you could try to download the well data you manually downloaded from BROLoket automatically using hydropandas:

```python
import hydropandas as hpd

# download a single head observation using the groundwater monitoring ID
o = hpd.GroundwaterObs.from_bro("GMW000000041261", 1)

# or download all head observations within an extent
oc = hpd.read_bro(extent=(xmin, xmax, ymin, ymax))
```

Hope that helps,

Davíd

**Reply by ArtemisRo** [reply]
Dear David,
Thanks for your reply. I have an ObsCollection now with the below code:
oc = hpd.read_bro(extent=(117850, 118180, 439550, 439900), keep_all_obs=False)
oc
with this code: oc.describe(), I can find the information for all wells. 
could you please let me know how to do it for indivitual object in this collection?  
Thanks in advance

**Reply by dbrakenhoff** [reply]
The individual Observation objects are stored in the "obs" column in the ObsCollection. You can extract individual Observation objects from the ObsCollection like this:

```python
o = oc.loc["<name>", "obs"]  # OR
o = oc["obs"].iloc[0]

o.describe()
```

**Comment by ArtemisRo** [comment]
Dear david,
I want to plot the grounwater level of some wells in the ObsCollection in the same plot with the below code, but I can not get the true result. I would be grateful if You could help me. 
Moreover I need a table for these wells with the same characteristics like (groundlevel, tube top, screen_top, screen_bottom, count, max, 75-percentage, mean, 25-percentage and min. May I ask you to help me with providing the code?
Thanks alot.

oc = hpd.read_bro(extent=(93000, 95000, 444000, 446000))
pstore = oc.to_pastastore(pstore_name='Krimpen')
oseries = pstore.oseries
ax = pstore.maps.oseries(backgroundmap=True)
for oseries_id in ['GLD000000025074', 'GLD000000025108', 'GLD000000025086']:
    oseries_data = oseries.get(oseries_id)
    ax = oseries_data.plot(ax=ax, label=oseries_id)

**Comment by dbrakenhoff** [comment]
Hi @rt84ro,

I recommend you just try to extract the data you require from the DataFrames and use `ax.plot()` to plot that data. `ObsCollection` and `pstore.oseries` are both essentially pandas.DataFrames. For manipulating/plotting and combining data from DataFrames I recommend you take a look at the [pandas documentation](https://pandas.pydata.org/docs/) or search StackOverflow for code snippets.

Hope that helps!

Davíd

**Reply by HMEUW** [reply]
See below some tips and trics for your problem.

Original code, inclusing pastastore.
The obs-column in the ObsCollection holds all observations, that column disappears in the pstore loop. I don't know why. I use ObsCollection in next part.

```import hydropandas as hpd
import hydropandas as hpd
import pastas as ps
import matplotlib.pyplot as plt

oc = hpd.read_bro(extent=(93000, 95000, 444000, 446000))
pstore = oc.to_pastastore(pstore_name='Krimpen')
oseries = pstore.oseries

print('obs in oc? ', 'obs' in oc.columns)
print('obs in oseries? ', 'obs' in oseries.columns)
```

Timeseries plot
```
fig, ax = plt.subplots()
for oseries_id in ['GLD000000025074', 'GLD000000025108', 'GLD000000025086']:
    o = oc.loc[oseries_id, 'obs']
    #o.plot(ax=ax, label=oseries_id)
    # used ax.plot instead of o.plot, because o.plot uses name of obs-column as legend item. 
    # That column has the name 'value', not so usefull in plot with 3 wells
    ax.plot(o.index, o['values'], label=oseries_id)
ax.legend()
```

Add data to oc, copy collection and add columns
```
oc_stats = oc.copy()
new_cols = ['obs_count', 'obs_max', 'obs_q75', 'obs_mean', 'obs_q25', 'obs_min']
oc_stats[new_cols] = None
for index, row in oc_stats.iterrows():
    # loop over all wells
    o = oc.loc[index, 'obs']
    oc_stats.loc[index, 'obs_count'] = o['values'].count()
    oc_stats.loc[index, 'obs_max'] = o['values'].max()
    oc_stats.loc[index, 'obs_q75'] = o['values'].quantile(0.75)
    oc_stats.loc[index, 'obs_mean'] = o['values'].mean()
    oc_stats.loc[index, 'obs_q25'] = o['values'].quantile(0.25)
    oc_stats.loc[index, 'obs_min'] = o['values'].min()

# some magic HydroPandas    
for new_col in new_cols:
    oc_stats.add_meta_to_df(new_col)
```

Quick plot of all stats. Skipping first columns, and leaving out some others.

```
fig, ax = plt.subplots()

# quick plot of stats
for col in oc_stats.columns[7:]:
    if col not in ['metadata_available', 'count', 'obs']:
        ax.scatter(oc_stats.x, oc_stats[col], label=col)
        
ax.legend()
```

**Reply by ArtemisRo** [reply]
Dear Hendrik,
Thank you so much for your great help. It works perfectly.

**Reply by HMEUW** [reply]
Thanks.

Can you later share your complete Python analysis of the groundwater levels in Krimpen? This kind of examples can be usefull for future newbies in HydroPandas.
